editor/src/components/text-editor/text-editor.tsx

```diff
@@ -58,16 +58,66 @@ const entities = {
 
 // canvas → editor
 export function escapeHTML(s: string): string {
-  return (
-    s
-      // a trailing newline is added by contenteditable for multiline strings, so get rid of it
-      .replace(/\n$/, '')
-      // clean up angular braces
-      .replace('<', entities.lesserThan)
-      .replace('>', entities.greaterThan)
-      // restore br tags
-      .replace(/\n/g, '\n<br />')
-  )
+  const withoutNewLines = s
+    // a trailing newline is added by contenteditable for multiline strings, so get rid of it
+    .replace(/\n$/, '')
+
+  //encode < and > when necessary
+  const encoded = encodeHTMLWhenNotInJsCode(withoutNewLines)
+
+  // restore br tags
+  return encoded.replace(/\n/g, '\n<br />')
+}
+
+// This is a very basic function to separate the real text content and the JS content in curly brackets
+// We only want to html encode in text content, but not in js content
+// E.g. we don't want to encode the < in `{ 2>3 ? "foo" : "bar" }
+function encodeHTMLWhenNotInJsCode(s: string): string {
+  let result = ''
+  let parenCounter = 0
+  let inSingleQuotation = false
+  let inDoubleQuotation = false
+
+  const isInJSExpression = () => parenCounter > 0
+
+  for (let ch of s) {
+    let characterToPrint = ch
+    switch (ch) {
+      case '{':
+        if (!inSingleQuotation && !inDoubleQuotation) {
```

Comment:
I wonder if these could be extracted into a mini-state machine for maximum encapsulation (but for this PR this is perfect as well)